cpp/src/arrow/io/meson.build

```diff
@@ -0,0 +1,75 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+io_tests = ['buffered_test', 'compressed_test', 'file_test', 'memory_test']
+
+foreach io_test : io_tests
+    test_name = 'arrow-io-@0@'.format(io_test.replace('_', '-'))
+    exc = executable(
+        test_name,
+        sources: '@0@.cc'.format(io_test),
+        dependencies: [arrow_test_dep],
+        implicit_include_directories: false,
+    )
+    test(test_name, exc)
+endforeach
+
+if needs_hdfs
+    hdfs_incdir = '../../../thirdparty/hadoop/include'
```

Comment:
From a purity perspective I think it would be better if we used hdfs as a subproject, and could create a dependency for it. However, I wouldn't find the original source for the file `thirdparty/hadoop/include/hdfs.h`

Reply:
https://github.com/apache/hadoop/blob/trunk/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/include/hdfs/hdfs.h is the original.

Reply:
Thanks! Does there happen to be a canonical link to download just the hadoop-hdfs-native-client?

Reply:
Apache Hadoop source archive includes it:
https://www.apache.org/dyn/closer.cgi/hadoop/common/hadoop-3.4.1/hadoop-3.4.1-src.tar.gz?action=download

Reply:
I'm not sure that there is a way with the wrapdb to natively extract just the file we need from a collection of projects like that without a top level CMakeLists.txt or meson.build configuration. I'll have to research that further if its a blocker here

Reply:
I don't think this is a blocker.
We can keep using the vendored `hdfs.h`.

We loads `libhdfs.so` manually by `dlload()` not compile time `-lhdfs`. So we just need `hdfs.h`. If `hdfs.h` has many backward incompatible changes, we may want to use upstream `hdfs.h`. But it's very stable so far. So we can keep using the vendored `hdfs.h`.